### ISSUE TITLE
PBEM/PBF 'Play' button is never enabled

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#5856]
**How can the problem be recreated?**
1. Launch game
2. Select PBF/PBEM
3. Note, despite filling in all fields, that 'Play' button stays disabled

**When the problem is fixed, what should have happened instead?**
- Play button should be enabled when minimum conditions have been met
- We probably should update the play button tooltip when PBF/PBEM is entered to be specific about what should be filled in

**Which Engine Version are you using?** 
df5a2242e8fba95182fe8a9376da3db12b28a29c

**Screenshots** 
![Screenshot from 2020-01-19 16-38-32](https://user-images.githubusercontent.com/12397753/72691456-5f0a1800-3ada-11ea-957f-0a4094d3147f.png)

**Zipped Save Game**

**Comment by RoiEXLab** [comment]
Have you configured an email server in your engine preferences?
I wanted to add an information dialog some time ago, but never actually did it

**Comment by RoiEXLab** [comment]
Second thought: I think I wanted to make configuring the email setting in engine preferences optional, so that the engine would ask if no credentials are available.
Then there would be no reason to require the credentials in this screen in the first place.
I even tried this in the past, but failed at some point for some reason I forgot

**Comment by DanVanAtta** [comment]
Is this a convenience item added? How much of the email server config needs to be set before the play button could be enabled?

**Comment by RoiEXLab** [comment]
For email you need to configure a full SMTP config; Hostname, Port, security protocol, username password.
For PbF you just need to enter your username and password and specify the server you want to configure it for

**Comment by DanVanAtta** [comment]
@RoiEXLab is there more background I can read on this? Why is SMTP server config a new requirement? 

Settings window was not intended to necessarily be the only place where settings are. They should generally be available where/when needed. If for example, you say "don't show this dialog again", then settings is a good place to turn it back on. Secondarily, to remove options from menus and consolidate the settings that were only available from the main loading screen to those that were only available in game (just because they were on different windows/screens).

This ticket certainly, at least IMO, represents something that must be fixed before 2.0 can be released. Specificially:
- play button being disabled needs to state what needs to be done to enable it
- a smpt server should not be required
- convenience settings/configuration for PBEM shoudl be available from that screen (in addition to settings window)

If we go to this length, it may be easy/reasonable enough to fix the PBEM/PBF multi-screen and split them into their own screens.

**Comment by RoiEXLab** [comment]
@DanVanAtta The SMTP server is not a new requirement at all, it has always been a necessary requirement for PbEM.
So whenever the user clicks send, the email is actually sent on their behalf using their own email.
There isn't any centralised mechanism for this.

**Comment by RoiEXLab** [comment]
The only difference is that "currently" the setting moved to client settings, where the preferences are stored system wide, whereas previously we had a custom java preferences store and temporarily stored credentials as part of the savegame and relied on some logic to remove those credentials before serialization.

**Comment by DanVanAtta** [comment]
Well, I was not able to start a PBF game in 2.0. 
In 1.9, I only had to fill in the email fields on the same screen and I could start. IIRC, 1.9 had SMTP as an optional thing for convenience to auto-send emails. Without auto-send, it was up to you to attach and send a game.

**Comment by DanVanAtta** [comment]
Do you disagree with the 3 requirements that need implementation I stated above? If no, please state what you think should be done to resolve this issue @RoiEXLab ? Do you disagree there is a problem? I hope not, as I almost had to look into the code to know how to start a PBEM/PBF game, I expect we'd get support tickets about that pretty fast.

**Comment by RoiEXLab** [comment]
> play button being disabled needs to state what needs to be done to enable it

I might be mistaken, but I think there might be a tooltip when you hover over the play button. If this is not the case, I'd definitely agree.

> a smpt server should not be required

I'm actually surprised to hear that it wasn't always required, I'm pretty sure it was when I touched the code.
I wonder what this would change though: Without any SMTP config the only thing that differentiates a normal local game from PbEM without automated email is the occasional dice server api calls. So yeah, we could definitely add a third "no automated send mechanism" that only makes use of the dice server, nothing else, or we could allow local games to use dice servers.

> convenience settings/configuration for PBEM should be available from the PBEM screen (in addition to settings window)

Either this, or we add a popup whenever the player tries to post a savegame, prompting the player to enter the credentials with the option to remember them which stores the setting in the client settings.

> I expect we'd get support tickets about that pretty fast

We might get them eventually, but PbEM is even less popular than PbF, which itself isn't that popular. So I don't think people will bother for quite some time.

**Comment by DanVanAtta** [comment]
> I might be mistaken, but I think there might be a tooltip when you hover over the play button. If this is not the case, I'd definitely agree.

The tooltip is generic / incorrect. It assumes you are on a game staging screen proper, ie: single or multiplayer side selection.

> I'm actually surprised to hear that it wasn't always required, I'm pretty sure it was when I touched the code.

No, PBEM without SMTP server integration AFAIK was more standard, and PBEM was esssentially just a way to get confirmed dice rolls.

> Without any SMTP config the only thing that differentiates a normal local game from PbEM without automated email is the occasional dice server api calls.

That's essentially what it is, it typically is a 'confirmed dice' mode. The impact is a few fold:
- users are going to get confused, they may have no idea what a SMTP server is
- I'm personally locked out. My gmail password is the most sensitive password I have. I do not want it persisted on any system.

> We might get them eventually, but PbEM is even less popular than PbF, which itself isn't that popular. So I don't think people will bother for quite some time.

We do not have the data, but that is not my impression. AFAIK some lobby games go on to finish over PBEM, that PBEM is more popular than PBF even. PBF, my impression, is more used for tourney when you want the save game to be public and not simply attached in emails.

Regardless, PBEM was essentially the way to get verified dice rolls, IMO a pretty significant use-case.

**Comment by panther2** [comment]
> Regardless, PBEM was essentially the way to get verified dice rolls, IMO a pretty significant use-case.
@DanVanAtta 
IMHO the way to get verified dice rolls is to register with the dice server only. That is totally independent from PBEM. PBEM IMHO is an alternative to PBF. Many PBF players switch to PBEM in case the forum is not available for whatever reason. Also players who don't want to play "publicly" on a forum use PBEM, then.

**Comment by panther2** [comment]
> We might get them eventually, but PbEM is even less popular than PbF, which itself isn't that popular.

At least PBF is most popular on www.axisandallies.org .

**Comment by DanVanAtta** [comment]
The only way to play with a dice server, was with either PBEM or PBF. Am I wrong there and there is some other way to enable the dice server for a TripleA game?

![Screenshot from 2020-01-24 00-43-25](https://user-images.githubusercontent.com/12397753/73055903-9f52f880-3e42-11ea-846d-7cc2bd221e18.png)

I mentioned we also do not have the numbers to know how many players use PBEM. I know there is a live-game use-case where players continue via PBEM and typically never PBF (unless it's a tourney).

@panther2 this is not a popularity contest, it does not matter relative numbers if a major use-case is broken (even if a minority). I'm interested to know the independent way to use verified dice without PBEM or PBF.

**Comment by DanVanAtta** [comment]
If you click on PBEM/PBF, you used to get this screen:

![Screenshot from 2020-01-24 00-46-57](https://user-images.githubusercontent.com/12397753/73056098-0c668e00-3e43-11ea-91ff-1b751e39014a.png)

It's getting to semantics now, not enabling the SMTP server generally assumed that the attachment of save games was manual, and it was still a PBEM game. After-all too, the button to get to the above screen is "PBEM/PBF", hence that association. Beyond that, verified dice for an offline game are pretty useless if the game is not shared somehow..

**Comment by DanVanAtta** [comment]
We should stay focused on what/how to fix the problem noted in this issue, this thread is getting overly long. Is there disagreement on the premise of the problem? As far as I'm concerned, the fact I had to nearly look at the code to learn how to enable the 'play' button indicates the UX is broken. 

Second, a SMTP server should not be needed for PBF, I was not able to find a way where filling in all PBF fields would enable the play button.

**Comment by RoiEXLab** [comment]
@DanVanAtta For PbF you have to configure your forum credentials in the engine preferences

**Comment by DanVanAtta** [comment]
That makes sense to be required.

If a PBEM/PBF game is started, is there a reason to change the connection/email configuration mid-game? I'm wondering if we should have those values as a setting window option.

**Comment by RoiEXLab** [comment]
I don't think this is a common thing people would want to do, but the current system definitely allows for this:
When re-loading a savegame you should be able to choose different settings, or load it as a local game without any "play by x" functionality

**Comment by DanVanAtta** [comment]
> When re-loading a savegame you should be able to choose different settings, or load it as a local game without any "play by x" functionality

AFAIK this has not changed.

---------

Sounds like we need to:
- remove PBF/PBEM settings from preferences window
- move PBF/PBEM settings to PBF/PBEM setup window
- make PBEM SMTP configuration optional
- update 'play' button tooltip text to be accurate and ensure the toottip reverts if a user clicks cancel.
- consider splitting PBEM/PBF to be two different screens activated by different buttons on the main game screen

**Comment by RoiEXLab** [comment]
@DanVanAtta 
My opinion:
1. Maybe. (See 2.)
2. That, or just implement a functionality that redirects to the engine preferences.
3. Sure.
4. Definitely
5. Mixed opinion on that one. They definitely share a lot of things that make this menu "an obvious step from a code perspective", but I'd actually cosider a completely different approach: To me the whole dice-server/PbF/PbEM thing seems like drop-in functionality which simply expands a normal local game, so adding a popup button in the local game screen to add additional "features" or swap out the dice server seems like a "logical" step to meß
Dunno, something to consider at least.

**Comment by DanVanAtta** [comment]
Don't offhand agree to redirect to engine preferences. Engine prefs is meant to be a secondary location,  redundant to the primary location where the options are relevant:
- https://learnui.design/blog/4-rules-intuitive-ux.html#1-obey-the-law-of-locality
- https://learnui.design/blog/the-3-laws-of-locality.html

Otherwise, the options in engine preferences should be really seldomly used and potentially global. IE: preferences being set from the launch screen, think of the 'screen resolution' setting in other games.

Re: 5) Making the verified dice option to be in-game IMO is a good move, but not mutually exlcusive. I think part of the issue in loading screens is we are overloading the PBEM/PBF to have two functions. Even if there is overlap, they are different enough and distinct that you don't really want a radio button to select between them. It's confusing for example if you select both PBEM/PBF tabs and fill everything out, which variant do you get?
This violated a UX principle to have a 1:1 correspondence between controls and actions. We have 1 button to enable 2 different functionalities.

- https://gcdtech.com/blog/the-7-principles-of-doet-and-how-to-simplify-any-task/ 
  - "ii) Affordances"
  - "4. Get the mappings right"
  - "v) Constraints "


So in other words, if you want to do a PBF game, you should not be allowed or even see unnecessary PBEM settings, they only cause confusion and make the flow to start a PBF more complex. Essentially, if you want to start a PBF, you should be presented with only that which you need to start a PBF and not extraneous (and unused) option.

**Comment by DanVanAtta** [comment]
Why do you think  this can be closed @RoiEXLab ? (https://github.com/triplea-game/triplea/issues/5025#issuecomment-590107267)

When I tested PBF/PBEM, I was left completely confused as how to start the game, so much so it seemed like it was broken. There is *nothing* to tell a user they need to go to a *completely* different screen to setup email server and credentials.

**Comment by RoiEXLab** [comment]
@DanVanAtta Didn't you add the tooltips for this button?
While I agree this is not the 100% ideal solution, but I thought this solution was good enough for now to release 2.0

**Comment by DanVanAtta** [comment]
The play button tooltip is still generic and has not been changed. Relying on a tooltip to tell a user to go to a different screen is better than nothing, but is still pretty broken UX. We have the required fields highlighted in red, yet we have effectively hidden fields (that live on another screen) that are required, making it all very inconsistent.

I'm working on moving the password and username field to the actual PBF/PBEM screens.

**Comment by RoiEXLab** [comment]
Ah ok, in this case my memory was incorrect, thanks for clarifying this. 👍

**Comment by DanVanAtta** [comment]
Remaining to complete this is to add the username and password fields to the PBEM & PBF screens. It should be straight forward. If we decide to not store credentials, then we'd just need to store username and build in a prompt for password when needed.

Still assigned to me, but if in the short term there is bandwidth, feel free to comment and take this over. @ron-murhammer @RoiEXLab

**Comment by DanVanAtta** [comment]
Closing, you can now enable the 'play' button on either PBF or PBEM by filling in just the forms available on those two respsective screens.
